src/abstractdataframe/sort.jl

```diff
@@ -682,16 +701,117 @@ function Base.sort!(df::AbstractDataFrame, cols=All();
                     lt::Union{Function, AbstractVector{<:Function}}=isless,
                     by::Union{Function, AbstractVector{<:Function}}=identity,
                     rev::Union{Bool, AbstractVector{Bool}}=false,
-                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward)
+                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward,
+                    checkunique::Bool=false)
 
     # exclude AbstractVector as in that case cols can contain order(...) clauses
     if cols isa MultiColumnIndex && !(cols isa AbstractVector)
         cols = index(df)[cols]
     end
     ord = ordering(df, cols, lt, by, rev, order)
     _alg = Sort.defalg(df, ord; alg=alg, cols=cols)
+    checkunique && _perform_uniqueness_checks(df, cols, lt, by, order)
```

Comment:
Is this check needed? We call `sort!` below which triggers uniqueness checks. Right?

Reply:
True, enabling `sort!` to do the checks made them redundant in the calling function. I'll remove them

Reply:
still pending to be removed - right?

Reply:
Right, sorry. It slipped my mind
Next commit won't have it

Reply:
@bkamins I was working on this, and it turned out to be a little bit more complex than I anticipated. The situation is...

If I remove the call to `_perform_uniqueness_checks` and let the other method (signature `sort!(df, algorithm, ordering, checkunique)`) handle it, the information about the columns is lost. The argument `cols` is used to create an "ordering" which can be of type `Perm` or `DFPerm`. There is already functionality to check _complexity_ of those, but their _uniqueness_ needs more context to be checked. Namely, we need the `cols` information which is either lost or just becomes really inconvenient to access. 

In the case where a `Perm` object is returned as `ord`, we cannot retrieve which column was selected from the `Perm` object.

At the moment I can't only think of 2 possible solutions:
1. Leave the "duplicate" uniqueness checks in both `sort!(df,...)` methods
2. Only check for uniqueness in `sort!(df, a::Algorithm, o::Ordering)` and overhaul all of the `order(...)` methods (several) to enable them to carry out the uniqueness checks there. 

Option 2 is perhaps the most elegant as we only have to check if `checkunique` is `true` once, but it would require the change of many other methods used all along the module, which could introduce unexpected behaviour. What would you advice?

Reply:
I am OK to keep "duplicate" in this PR in this case. I understand that this check is not expensive anyway - right?

Reply:
To the best of my knowledge, the most expensive part is the actual check for duplicate elements, i.e the `allunique` function. The code to detect complex order clauses is fairly simple and doesn't require much memory allocation, other than keeping column indexes. And since `allunique` is based on the hashing algorithm used for `groupby`, I assume its more than performant enough

Reply:
I just had another idea that would spare the second call to `_perform_uniqueness_checks`. When `sort!(df, Algorithm, Ordering)` is called from `sort!(df, cols, ...)` we can set `checkunique` to false, since the uniqueness checks were already made and at that point, either we threw an error because there were duplicates, or we didn't and we don't have to check once more. The kwarg can still be used and set to `true` by the user when calling it directly, which is the expected behaviour

Reply:
Yes - this makes sense.